### PR TITLE
Make eval work faster

### PR DESCRIPTION
**Pull request by NickleDave**

### CODE REVIEW

**Comment by NickleDave** [comment]
@yardencsGitHub I have been working on this feature branch to get eval / validation working the same way that predict does, by converting a spectrogram into a batch of consecutive windows, where the step size = window size instead of a step size of 1 as in the training sets

the problem is that it's giving me very low accuracy on the validation set. I find the same issue with canary and BF song

Can you take a look at it?
You will probably find it faster to test / debug using BF data, e.g gy6or6 from the BF Song Repository (The fact that it's still a really long start up time for canary data is a separate issue we need to look at but let's work on this first)
It could be that it's actually computing things correctly but there's something different about the network as implemented in torch. Not sure.

**Comment by NickleDave** [comment]
I'm also seeing a bug where it doesn't stop training even though it says that it will because > patience number of validations have gone by without accuracy increasing

I think I know why that is but can't fix right now. Commenting here so I don't forget

**Comment by NickleDave** [comment]
Here are the configs I've been using to test on "typical" sized datasets
[configs.tar.gz](https://github.com/NickleDave/vak/files/4481613/configs.tar.gz)

**Comment by NickleDave** [comment]
Another possible gotcha is that we change from `model.train()` to `model.eval()` mode inside the `_eval` loop -- maybe there's something like dropout that's on by default we're not accounting for?
I'd be surprised if that makes such a difference but just trying to think of possible sources of error

**Comment by yardencsGitHub** [comment]
I have a dataset of 815 audio files from gy6or6.
Previously I converted this dataset to my own matlab-based format so I will try to start with that.

I really hope no big issues happened by the move to pytorch.
I thought one of the first things to check was that the results of training are very similar (even if converting saved models was impractical)

**Comment by NickleDave** [comment]
It will be easier if you just use the gy6or6 data off the repository
https://figshare.com/articles/Bengalese_Finch_song_repository/4805749

This is a good opportunity to deeply understand the algorithm; the results should be independent of framework and if not then it's our job to know why

**Comment by yardencsGitHub** [comment]
Ok. I think I get the same results.
Validation error does not improve and the training doesn't stop.
I'll look into it

**Comment by NickleDave** [comment]
I think it's a relatively simple issue with the new `eval` loop; something about calling `reshape` on the network output changes the data, which is what causes it to be "wrong".
But it's not clear to me yet how or why.

**Comment by NickleDave** [comment]
I was able to verify with pdb for one batch that I found a way to reshape the output of the network without changing the data, and the accuracy was much higher at the first validation point, but then it keeps going down and the loss keeps going up on the validation step, even though loss during training is decreasing. Will look at it again in the AM

**Comment by NickleDave** [comment]
Another possibility might be that the way I'm reshaping the spectrograms into stacks of windows is not preserving the data correctly so that it breaks the relationship between x and y.

The way to test this would be to compare whether
`torch.eq(x[0] == spect[:, :n_time_bins_per_window])`
`torch.eq(x[1] == spect[n_time_bins_per_window, :n_time_bins_per_window*2])`
and so on

Having functioning tests in place would have helped avoid problems like this

**Comment by yardencsGitHub** [comment]
The force pushing (I know you like the github nomenclature) placed 'tracking' text in model.py that prevents running 'train'.

**Comment by yardencsGitHub** [comment]
to follow the above: https://github.com/NickleDave/vak/pull/163#issuecomment-614683831
How do I locate the correct spect to load. (Unlike my matlab files, here 'prep' created a folder with spectrogram files)
How are they loaded (loblib, pickle and the like or pandas?)

**Comment by NickleDave** [comment]
> The force pushing (I know you like the github nomenclature) placed 'tracking' text in model.py that prevents running 'train'. I removed those and got a strange size for the data tensor (68,1,257,88).
> What is source for the 68?

Can you clarify what you mean by 'tracking'? Do you mean I accidentally left a  `pdb.set_trace()` in the code or something? Is there a specific error you get when you run `train`?

68 is the batch size.
The transform applied to the spectrogram is to pad so that it can be reshaped into a stack of windows, and then reshape.
So the batch size will be the number of windows you get after padding and taking steps with size equal to the window.
If you look at `vak.transforms.defaults.EvalItemTransform` you'll see how it's performed.

`.npz` files are sort of the `numpy` equivalent of `.mat` files, they can be opened with `numpy.load`

Probably the quickest way to get the spect before the transform is to insert a breakpoint just before the main loop in `_eval`, and use `n` (for next) in the Python debugger to advance to the line *just before* unpacking a batch from the `dataloader`, then use `s` to step into the dataloader.__get_item__ method, see what the `index` is (probably `0`?) and use that to index into `spect_paths`, which is what `__get_item__` does.
Then go back to using `n` to advance until the `__get_item__` returns `item` and you pop back up a frame to inside the `_eval` method, where you will now have `x` present in the namespace.
Use the index you obtained inside the `__get_item__` method to get the spectrogram out of the file:
```
spect_dict = np.load(self.spect_paths[index])
spect = spect_dict[self.spect_key]
```
then go on to do the comparison from above
```
torch.eq(x[0] == spect[:, :n_time_bins_per_window])
torch.eq(x[1] == spect[n_time_bins_per_window, :n_time_bins_per_window*2])
```

**Comment by yardencsGitHub** [comment]
The data loading is with a multiworker process.. not easy to figure out the actual get_item step

**Comment by NickleDave** [comment]
Ah sorry if I confused you.

The `__get_item__` method is not called explicitly in the `vak.Model._eval` method.
I just meant that the `dataloader`, which in this case is called `eval_data`, implicity calls `__get_item__` itself on its `dataset` attribute, so if you just keep typing `n` in the debugger you will find yourself eventually inside the `__get_item__` method. Agreed that its confusing if you try to read the `torch.dataloader` source code and figure out what's going on with all the multiprocessing stuff. But ultimately one of those processes has to index into the dataset, which is what triggers the `__get_item__` method call.

I'm back trying to debug now

**Comment by NickleDave** [comment]
Turns out my idea for figuring out which spect it is doesn't work anyway :(

But yes I am now convinced that `transforms.functional.reshape_to_transform` is the problem; reshaping does not preserve the ordering of the data as I thought.

Calling the `reshape` method on a Numpy array essentially ravels the array using the specified indexing order (default is C order, "row major") and then builds a new array with the desired shape using the same indexing order.
https://docs.scipy.org/doc/numpy/reference/generated/numpy.ravel.html

A simple demonstration of why this breaks.
Imagine this is our spectrogram:  
`>>> spect = np.asarray([[0, 0, 0], [1, 1, 1]])`
and we attempt to turn it into a batch of windows with width 1 using `reshape` like so:
`>>> spect_batch = spect.reshape(-1, 2, 1)`
We end up with:  
```
>>> print(spect_batch)
array([[[0],
        [0]],
       [[0],
        [1]],
       [[1],
        [1]]])
```
i.e. our first window is 
```
array([[0],
       [0]])
```
when we expect it to be
```
array([[0],
       [1]])
```

I think the fix is to use `numpy.lib.stride_tricks.as_strided`, something like they do in `skimage.util.view_as_windows`:
https://github.com/scikit-image/scikit-image/blob/f1b7cf60fb80822849129cb76269b75b8ef18db1/skimage/util/shape.py#L98

**Comment by NickleDave** [comment]
Actually I think the answer was similar to `skimage.util.view_as_block` (where "block" = non-overlapping windows)
https://github.com/scikit-image/scikit-image/blob/f1b7cf60fb80822849129cb76269b75b8ef18db1/skimage/util/shape.py#L9
I tested "by hand" that this didn't mess up the spectrogram and have re-written the function, will finish tidying up, force push, and then see if it fixes everything in the AM
